cli/internal/cmd/apply.go

```diff
@@ -0,0 +1,505 @@
+		initRequired = true && !a.flags.skipPhases.contains(skipInitPhase)
+	}
+	a.log.Debugf("Init RPC required: %t", initRequired)
+
+	// Validate input arguments
+
+	// Validate Kubernetes version as set in the user's config
+	// If we need to run the init RPC, the version has to be valid
+	// Otherwise, we are able to use an outdated version, meaning we skip the K8s upgrade
+	a.log.Debugf("Validating Kubernetes version %s", conf.KubernetesVersion)
+	validVersion, err := versions.NewValidK8sVersion(string(conf.KubernetesVersion), true)
+	if err != nil {
+		a.log.Debugf("Kubernetes version not valid: %s", err)
+		if initRequired {
+			return err
+		}
+		a.log.Debugf("Checking if user wants to continue anyway")
+		if !a.flags.yes {
+			confirmed, err := askToConfirm(cmd,
+				fmt.Sprintf(
+					"WARNING: The Kubernetes patch version %s is not supported. If you continue, Kubernetes upgrades will be skipped. Do you want to continue anyway?",
+					validVersion,
+				),
+			)
+			if err != nil {
+				return fmt.Errorf("asking for confirmation: %w", err)
+			}
+			if !confirmed {
+				return fmt.Errorf("aborted by user")
+			}
+		}
+		a.flags.skipPhases = append(a.flags.skipPhases, skipK8sPhase)
+		a.log.Debugf("Outdated Kubernetes version accepted, Kubernetes upgrade will be skipped")
+	}
+	if versions.IsPreviewK8sVersion(validVersion) {
+		cmd.PrintErrf("Warning: Constellation with Kubernetes %s is still in preview. Use only for evaluation purposes.\n", validVersion)
+	}
+	conf.KubernetesVersion = validVersion
+	a.log.Debugf("Target Kubernetes version set to %s", conf.KubernetesVersion)
+
+	// Validate microservice version (helm versions) in the user's config matches the version of the CLI
+	// This makes sure we catch potential errors early, not just after we already ran Terraform migrations or the init RPC
+	if !a.flags.force {
+		if err := validateCLIandConstellationVersionAreEqual(constants.BinaryVersion(), conf.Image, conf.MicroserviceVersion); err != nil {
+			return err
+		}
+	}
+
+	// Constellation on QEMU or OpenStack don't support upgrades
+	// If using one of those providers, make sure the command is only used to initialize a cluster
+	if !(conf.GetProvider() == cloudprovider.AWS || conf.GetProvider() == cloudprovider.Azure || conf.GetProvider() == cloudprovider.GCP) {
+		if !initRequired {
+			return fmt.Errorf("upgrades are not supported for provider %s", conf.GetProvider())
+		}
+		// Skip Terraform phase
+		a.log.Debugf("Skipping Infrastructure upgrade")
+		a.flags.skipPhases = append(a.flags.skipPhases, skipInfrastructurePhase)
+	}
+
+	// Check if Terraform state exists
+	tfStateExists, err := a.tfStateExists()
+	if err != nil {
+		return fmt.Errorf("checking Terraform state: %w", err)
+	}
+	if !tfStateExists {
+		a.log.Debugf("No Terraform state found in current working directory. Assuming self-managed infrastructure. Infrastructure upgrades will not be performed.")
```

Comment:
nit: I'd translate the tfStateExists into skipInfraStructurePhase and make the variable local in the if clause